### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by brenthuisman**

### CODE REVIEW

File: CONTRIBUTING.md
Comment (noraabiakar) [inline CONTRIBUTING.md:11]: There's an incomplete sentence "Please see the".
Also can we split off the lines? They look like they may be too long.

Comment (brenthuisman) [reply CONTRIBUTING.md]: Done!

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:13]: Do we have a template? The usual procedure seems to be to require a small explanation of what the user did, what what happend, and what was the desired outcome (instead).

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:21]: Our policy so far was the exact opposite.

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:37]: Could remove 'feature', reads clearer, since it's not always features. Also: less jargon.

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:45]: Mention `bors try`?
Also, make clear that _we_ will merge their patch.
Finally: The copyright form?

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:69]: This should be moved to the previous section.

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:76]: This is not public?

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:57]: Run unit tests, incl bors.

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:60]: If adding/changing the public C++ API, provide python wrappings.

File: CONTRIBUTING.md
Comment (thorstenhater) [inline CONTRIBUTING.md:49]: No guidelines are state below (explicitely as such)

**Review by thorstenhater — CHANGES_REQUESTED** [review]
A few comments. But I think we should have _all_ current core contributors sign off on this.

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:13]: We have [issue templates](https://github.com/arbor-sim/arbor/tree/master/.github/ISSUE_TEMPLATE), based on the Github provided templates.

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:21]: On august 10 I announced the change on Slack: "Github just accepted the request for 'Discussions'. I've placed a first post with some explanation, as well as two Issue templates, meant to subtly guide usage-related questions to 'Discussions'. In case someone is bothered by these Issue templates, I can remove them and set it to a blank template (to keep the message about usage-related questions in place)."

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:37]: Done!

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:76]: It isn't (because Slack always requires an account), but AFAIK anybody requesting access gets it. And I think people should get in if they want.

But, it's a good idea to emphasize that Discussions is more for user related questions, and Slack for developers, so I've added a hint.

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:45]: What does `bors try` do that Travis doesn't?

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:69]: OK

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:57]: We rely on Travis which is automatic for most PRs. See also previous question about bors.

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:60]: Done.

File: CONTRIBUTING.md
Comment (brenthuisman) [reply CONTRIBUTING.md:49]: I don't follow, the list that follows seems accurately describes as guidelines to me.

File: CONTRIBUTING.md
Comment (thorstenhater) [reply CONTRIBUTING.md:21]: Thanks for the reminder.

File: CONTRIBUTING.md
Comment (thorstenhater) [reply CONTRIBUTING.md]: GPU related, mainly.

Comment (brenthuisman) [reply CONTRIBUTING.md]: I'll add a line saying they can use bor to run GPU tests.

File: CONTRIBUTING.md
Comment (noraabiakar) [inline CONTRIBUTING.md:9]: Minor suggestion: maybe replace `ways to contribute, from writing tutorials or blog posts` with `ways to contribute: writing tutorials or blog posts`

File: CONTRIBUTING.md
Comment (noraabiakar) [inline CONTRIBUTING.md:75]: Another minor suggestion: `to help you get` instead of `to help getting`

File: CONTRIBUTING.md
Comment (noraabiakar) [inline CONTRIBUTING.md:82]: `If you add` instead of `If adding` fits better with the rest of the list.

File: CONTRIBUTING.md
Comment (noraabiakar) [reply CONTRIBUTING.md:69]: I would agree that the first point in this list is not actually a review guideline, but more of a tip? Maybe it can be added to `Some tips to help you get a positive review`

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (noraabiakar) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: Maybe this file needs to be referenced somewhere? Otherwise I don't see it being used. In general though, I don't know whether it is needed. All of the points listed here are listed in CONTRIBUTING.md

**Review by noraabiakar — CHANGES_REQUESTED** [review]
A few minor suggestions.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (brenthuisman) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: This file will be used to autopopulate a PR message in the Github UI. We can of course not use an automatic template for that, but given the reality of human beings probably not first reading CONTRIBUTING.md, I thought it would be helpful to have a quick tickable list to nudge people into doing the things we like to see before a PR is merged.

Comment (brenthuisman) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: I can't reply directly to the open comment above, but it's in the commit I'm about to push.

File: CONTRIBUTING.md
Comment (noraabiakar) [inline CONTRIBUTING.md:69]: I actually meant to keep this line the same as before: `Each pull request is reviewed according to these guidelines:`
but to add: `You can use labels to categorize your PR. For very large PRs, it is likely the reviewer will have
many questions and comments. It can be helpful to reach out on our Slack, Discussions or by email
beforehand if you've big things planned for Arbor!` to the list of tips right underneath.

Comment (brenthuisman) [reply CONTRIBUTING.md:69]: Wouldn't that put it back to what it was? I responded to your point about adding `Some tips to help you get a positive review`, or do you mean to make a flat list out of the section?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (bcumming) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: I don't think that we need this.
It adds a lot of boilerplate that we haven't needed to date.
I think that a better requirement is to ensure that an issue has been created that clearly states the work, and that the PR description describes what has been done.

Our policy is that the PR description will be used as the comment on the final squashed commit. PRs often change, with features removed or added over the course of the review, and the PR description should be updated to reflect the changes.

**Review by bcumming — CHANGES_REQUESTED** [review]
I have one request for change. To date we have not needed a PR template, and we have a well established policy for PR descriptions, which should be added to the CONTRIBUTING guide, and enforced during reviews.

File: CONTRIBUTING.md
Comment (noraabiakar) [reply CONTRIBUTING.md:69]: In lines 74-75 you already had `Some tips to help you get a positive review`. I was just suggesting to move lines 71-73 till after line 75 and reword them so they're a tip. And keep line 69 as: `Each pull request is reviewed according to these guidelines:`

Comment (brenthuisman) [reply CONTRIBUTING.md:69]: OK!

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (brenthuisman) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: This works well if there has been out of band (Slack) communication about the PR, but maybe not well if someone wants to provide a bit more motivation and context than a dry change summary. I've added a "change summary" section to the template, and an instruction that this should be according to current policy. Upon merge, just this section can be used for the commit message.

Comment (bcumming) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: My main concern is that our current method is well established, and there does not appear to be a strong reason for making the change. Templates are usually added to solve a problem (e.g. we observe that PRs have incomplete or incorrect messages), and we should wait until we observe problems that need solving, then discuss solutions like templates.

The current approach allows committers to add context. The issue associated with the PR can also provide context.

Comment (brenthuisman) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: The strong reason is that external contributors have no idea about our methods. I have added a line to Contributing.md, because you are correct to point out it was not there.

The idea is that it helps us by helping contributors discover CONTRIBUTING exists, and layout the PR message. I've condensed the template slightly: what we care about is motivation and the change summary, and it's helpful that they are separated. We could condense the checklist maybe, but I'd like to keep the link to CONTRIBUTING somewhere.

File: CONTRIBUTING.md
Comment (thorstenhater) [reply CONTRIBUTING.md:69]: Commit logical units with clear commit messages.

The commits might the wrong criterion. The whole PR should be a well defined unit.

Comment (brenthuisman) [reply CONTRIBUTING.md:69]: See l77:
`    * Add change summary to the PR message. If you remembered to commit logical units, this could simply be a bulleted list of all commit messages in the PR.`

Is this sufficient?

Comment (thorstenhater) [reply CONTRIBUTING.md:69]: 👍

**Review by thorstenhater — APPROVED** [review]

**Review by noraabiakar — APPROVED** [review]
I would still prefer to not have a template for pull requests, but I'll leave that up to you and Ben.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (bcumming) [reply .github/PULL_REQUEST_TEMPLATE.md:1]: We don't have external collaborators knocking our doors down to make PRs, so we don't know what the problems are with their PRs. So how about spelling out our requirements for contributing in the `CONTRIBUTIONS.md` file, and if we are lucky enough to have external contributions, they can be pointed there if their PR isn't up to scratch.

**Comment by brenthuisman** [comment]
CONTRIBUTING is now more or less a pointer to the wiki. I've left a welcome message and contact info in place.

**Review by bcumming — APPROVED** [review]
Looks good!
Thanks. Let's see how we go with a "light touch" contributions policy, and adjust it as issues arise.

**Pull request merged**
